model_hub/model_hub/huggingface/_trial.py

```diff
@@ -0,0 +1,245 @@
+                split=f"train[:{data_config.validation_split_percentage}%]",
+            )
+            datasets["train"] = hf_datasets.load_dataset(
+                data_config.dataset_name,
+                data_config.dataset_config_name,
+                split=f"train[{data_config.validation_split_percentage}%:]",
+            )
+    else:
+        data_files = {}
+        if data_config.train_file is not None:
+            data_files["train"] = data_config.train_file
+        if data_config["validation_file"] is not None:
+            data_files["validation"] = data_config.validation_file
+        extension = data_config.train_file.split(".")[-1]
+        if extension == "txt":
+            extension = "text"
+        datasets = hf_datasets.load_dataset(extension, data_files=data_files)
+    return datasets
+
+
+class BaseTransformerTrial(PyTorchTrial):
+    """
+    This is the base trial class for transformers with a default init and train_batch method.
+
+    You can subclass BaseTransformerTrial to customize a trial for your own usage by filing in
+    the expected methods for data loading and evaluation.
+
+    See examples/huggingface/token-classification/ner_trial.py for an example.
+    """
+
+    def __init__(self, context: PyTorchTrialContext) -> None:
+        self.context = context
+        self.hparams = AttrDict(context.get_hparams())
+        self.data_config = AttrDict(context.get_data_config())
+
+        # Parse hparams and data_config.
+        (
+            config_args,
+            tokenizer_args,
+            model_args,
+        ) = hf_parse.default_parse_config_tokenizer_model_args(self.hparams)
+        optimizer_args, scheduler_args = hf_parse.default_parse_optimizer_lr_scheduler_args(
+            self.hparams
+        )
+
+        self.config, self.tokenizer, self.model = build_using_auto(
+            config_args, tokenizer_args, self.hparams.model_mode, model_args
+        )
+        self.model = self.context.wrap_model(self.model)
+
+        self.optimizer = self.context.wrap_optimizer(
+            build_default_optimizer(self.model, optimizer_args)
+        )
+
+        if self.hparams.use_apex_amp:
+            self.model, self.optimizer = self.context.configure_apex_amp(
+                models=self.model,
+                optimizers=self.optimizer,
+            )
+
+        self.lr_scheduler = self.context.wrap_lr_scheduler(
+            build_default_lr_scheduler(self.optimizer, scheduler_args),
+            LRScheduler.StepMode.STEP_EVERY_BATCH,
+        )
+        self.grad_clip_fn = (
+            lambda x: torch.nn.utils.clip_grad_norm_(x, optimizer_args.max_grad_norm)
+            if optimizer_args.max_grad_norm > 0
+            else None
+        )
+
+    def train_batch(self, batch: Any, epoch_idx: int, batch_idx: int) -> Any:
+        # By default, all HF models return the loss in the first element.
+        outputs = self.model(**batch)
+        loss = outputs[0]
```

Comment:
are there any HF models which return a bare loss, rather than a list of outputs?  Like, is that a case we need to check for?

Reply:
It is by default (see https://github.com/huggingface/transformers/blob/v4.3.3/src/transformers/trainer.py#L1326) but I have added handling for dict as well.